fork-off/src/fetching.rs

```diff
@@ -1,202 +1,78 @@
-        BlockHash::new(&block)
-    }
+    async fn get_full_state_at_block(&self, block_hash: BlockHash, num_workers: u32) -> Storage {
+        info!("Fetching state at block {:?}", block_hash);
 
-    // The start_key is not included in the range. The result is `count` keys that appear after `start_key` in the
-    // lexicographic ordering of all keys in storage.
-    async fn get_keys_range(
-        &self,
-        count: u32,
-        start_key: Option<StorageKey>,
-        block_hash: Option<BlockHash>,
-    ) -> Vec<StorageKey> {
-        let prefix = StorageKey::new("");
-        let body = get_keys_paged_body(prefix, count, start_key, block_hash);
-        let keys: Vec<StorageKey> = self.make_request(body).await;
-        keys
-    }
-
-    async fn get_all_keys(&self, block_hash: BlockHash) -> Vec<StorageKey> {
-        let mut last_key_fetched = None;
-        let mut all_keys = Vec::new();
-        loop {
-            let new_keys = self
-                .get_keys_range(KEYS_BATCH_SIZE, last_key_fetched, Some(block_hash.clone()))
-                .await;
-            all_keys.extend_from_slice(&new_keys);
-            info!(
-                "Got {} new keys and have {} in total",
-                new_keys.len(),
-                all_keys.len()
-            );
-            if new_keys.len() < KEYS_BATCH_SIZE as usize {
-                break;
-            }
-            last_key_fetched = Some(new_keys.last().unwrap().clone());
-        }
-        all_keys
-    }
-
-    async fn get_value(&self, key: StorageKey, block_hash: BlockHash) -> StorageValue {
-        let body = get_storage_value_body(key, Some(block_hash));
-        self.make_request(body).await
-    }
-
-    async fn get_values(
-        &self,
-        keys: Vec<StorageKey>,
-        block_hash: BlockHash,
-        num_workers: u32,
-    ) -> Storage {
-        let n_keys = keys.len();
-        let input = Arc::new(Mutex::new(keys));
-        let output = Arc::new(Mutex::new(HashMap::with_capacity(n_keys)));
+        let (input, key_fetcher) = self.client.stream_all_keys(&block_hash);
+        let output = Arc::new(Mutex::new(HashMap::new()));
         let mut workers = Vec::new();
 
-        for id in 0..(num_workers as usize) {
-            workers.push(StateFetcher::value_fetching_worker(
-                self.http_rpc_endpoint.clone(),
-                id,
+        for _ in 0..(num_workers as usize) {
+            workers.push(self.value_fetching_worker(
                 block_hash.clone(),
                 input.clone(),
                 output.clone(),
             ));
         }
+
         info!("Started {} workers to download values.", workers.len());
-        join_all(workers).await;
-        assert!(input.lock().is_empty(), "Not all keys were fetched");
+        let (res, _) = join!(key_fetcher, join_all(workers));
+        res.unwrap();
+
         let mut guard = output.lock();
         std::mem::take(&mut guard)
```

Comment:
huh...those two lines can be simplified as `output.into_inner()`.

Reply:
That's a leftover from my code :P Good to know :)

Reply:
Hm, `output.into_inner()` gives `cannot move out of an `Arc``...